src/oauth.tsx

```diff
@@ -304,48 +317,129 @@ app.post("/token", cors(), async (c) => {
       );
     }
 
-    const token = await db.query.accessTokens.findFirst({
-      where: eq(accessTokens.code, form.code),
-      with: { application: true },
-    });
+    return await db
+      .transaction(
+        async (tx) => {
+          const accessGrantResult = await tx
+            .select()
+            .from(accessGrants)
+            .for("update")
+            .where(eq(accessGrants.token, accessGrantToken))
+            .limit(1);
 
-    if (token == null || token.grant_type !== "authorization_code") {
-      return c.json(
+          const accessGrant = accessGrantResult[0];
+
+          if (
+            accessGrant === undefined ||
+            accessGrant.applicationId !== application.id ||
+            accessGrant?.revokedAt !== null
+          ) {
+            return c.json(
+              {
+                error: "invalid_grant",
+                error_description: INVALID_GRANT_ERROR_DESCRIPTION,
+              },
+              400,
+            );
+          }
+
+          const notAfter =
+            accessGrant.createdAt.valueOf() + accessGrant.expiresIn;
+          if (Date.now() > notAfter) {
+            return c.json(
+              {
+                error: "invalid_grant",
+                error_description: INVALID_GRANT_ERROR_DESCRIPTION,
+              },
+              400,
+            );
+          }
+
+          if (accessGrant.redirectUri !== form.redirect_uri) {
+            return c.json(
+              {
+                error: "invalid_grant",
+                error_description: INVALID_GRANT_ERROR_DESCRIPTION,
+              },
+              400,
+            );
+          }
+
+          if (form.scope?.some((s) => !accessGrant.scopes.includes(s))) {
+            return c.json(
+              {
+                error: "invalid_scope",
+                error_description:
+                  "The requested scope is invalid, unknown, or malformed.",
+              },
+              400,
+            );
+          }
+
+          // Revoke the access grant:
+          await tx
+            .update(accessGrants)
+            .set({
+              revokedAt: new Date(),
+            })
+            .where(eq(accessGrants.token, accessGrantToken));
```

Comment:
Currently I'm setting `revokedAt` here, which means our database will have loads of now useless access grants in it. It would be a reasonable decision to just delete the access grant from the database in this case.

Reply:
@dahlia would like your thoughts here, a trade-off is that by not setting `revoked` and just deleting the access grant, you can't detect authorization code reuse attacks, where the authorization code is reused and consequently all created access tokens from that authorization code should be invalidated (grants, this would imply having some relationship between access grants and access tokens)